src/conn/conn_compact.c

```diff
@@ -21,21 +21,269 @@ __compact_server_run_chk(WT_SESSION_IMPL *session)
+ *     Pre-fill compact related statistics for the given file.
+ */
+int
+__wt_background_compact_start(WT_SESSION_IMPL *session)
+{
+    WT_BACKGROUND_COMPACT_STAT *compact_stat;
+    WT_BM *bm;
+    WT_DECL_RET;
+    uint32_t id;
+    const char *uri;
+
+    bm = S2BT(session)->bm;
+    id = S2BT(session)->id;
+    uri = bm->block->name;
+
+    compact_stat = __background_compact_get_stat(session, uri, id);
+
+    /* If the table is not in the list, allocate a new entry and insert it. */
+    if (compact_stat == NULL) {
+        WT_ERR(__wt_calloc_one(session, &compact_stat));
+        WT_ERR(__wt_strdup(session, uri, &compact_stat->uri));
+        compact_stat->id = id;
+        __background_compact_list_insert(session, compact_stat);
+    }
+
+    /* Fill starting information prior to running compaction. */
+    WT_ERR(bm->size(bm, session, &compact_stat->start_size));
+    compact_stat->prev_compact_time = __wt_clock(session);
+
+    return (0);
+
+err:
+    __wt_free(session, compact_stat);
+
+    return (ret);
+}
+
+/*
+ * __wt_background_compact_end --
+ *     Fill resulting compact statistics in the background compact tracking list for a given file.
+ */
+int
+__wt_background_compact_end(WT_SESSION_IMPL *session)
+{
+    WT_BACKGROUND_COMPACT_STAT *compact_stat;
+    WT_BM *bm;
+    WT_CONNECTION_IMPL *conn;
+    wt_off_t bytes_recovered;
+    int64_t id;
+    const char *uri;
+
+    bm = S2BT(session)->bm;
+    id = S2BT(session)->id;
+    uri = bm->block->name;
+
+    compact_stat = __background_compact_get_stat(session, uri, id);
+    conn = S2C(session);
+
+    WT_RET(bm->size(bm, session, &compact_stat->end_size));
+    compact_stat->bytes_rewritten = bm->block->compact_bytes_rewritten;
+    bytes_recovered = compact_stat->start_size - compact_stat->end_size;
+
+    /*
+     * If the file failed to decrease in size, mark as an unsuccessful attempt. It's possible for
+     * compaction to do work (rewriting bytes) while other operations cause the file to increase in
+     * size.
+     */
+    if (bytes_recovered <= 0) {
+        compact_stat->consecutive_unsuccessful_attempts++;
+        compact_stat->prev_compact_success = false;
+    } else {
+        compact_stat->consecutive_unsuccessful_attempts = 0;
+        conn->background_compact.files_compacted++;
+        compact_stat->prev_compact_success = true;
+
+        /*
+         * Update the moving average of bytes rewritten across each file compact attempt. A
+         * weighting of 10% means that we are effectively considering the last 10 attempts in the
+         * average.
+         */
+        conn->background_compact.bytes_rewritten_ema =
+          (uint64_t)(0.1 * bm->block->compact_bytes_rewritten +
+            0.9 * conn->background_compact.bytes_rewritten_ema);
+    }
+
+    return (0);
+}
+
+/*
+ * __background_compact_list_cleanup --
+ *     Free all or any entry that has not been updated for more than a day in the background compact
```

Comment:
We may not want to say "a day" here as it is now something that depends on which mode we are running. I would suggest something along those lines:
```
Free all or any entry that has been idle for too long ...
```

Reply:
But we should also consider that there is an argument. Should we mention that the cleanup can be requested too?

Reply:
Changed to 

> Free all entries as part of cleanup or any entry that has been idle for too long in the background compact tracking list.

https://github.com/wiredtiger/wiredtiger/pull/9541/commits/6f54c40e19073acbc673f92af440a2daefa5dee6